### ISSUE TITLE
chiang_hair_bsdf node compile error on MacOS

### ISSUE THREAD

**Issue by ashwinbhat** [AcademySoftwareFoundation/MaterialX#2056]
The MSL shaders generated for chiang_hair_bsdf node produce compilation errors on MacOS.

Steps: 
build MaterialX viewer on MacOS
load simple_hair_default.mtlx 

Errors:
program_source:1364:28: error: 'angles' declared as array of references of type 'float2 &'
        thread vec2& angles[4]
                           ^
program_source:1383:62: error: 'Ap' declared as array of references of type 'float3 &'
    void mx_hair_attenuation(float f, vec3 T, thread vec3& Ap[4])  // Ap
                                                             ^
program_source:1448:24: error: unexpected type name 'float3': expected expression
        vec3 tint[4] = vec3[](tint_R, tint_TT, tint_TRT, tint_TRT);
                       ^
program_source:7:14: note: expanded from macro 'vec3'
#define vec3 float3
             ^
program_source:1448:29: error: expected expression
        vec3 tint[4] = vec3[](tint_R, tint_TT, tint_TRT, tint_TRT);
                            ^
program_source:1453:22: error: unexpected type name 'float2': expected expression
        vec2 vs[4] = vec2[](roughness_R, roughness_TT, roughness_TRT, roughness_TRT);
                     ^
program_source:6:14: note: expanded from macro 'vec2'
#define vec2 float2
             ^
program_source:1453:27: error: expected expression
        vec2 vs[4] = vec2[](roughness_R, roughness_TT, roughness_TRT, roughness_TRT);
                          ^
program_source:1541:14: warning: unused variable 'dirAlbedo' [-Wunused-variable]
        vec3 dirAlbedo = mx_ggx_dir_albedo(NdotV, avgAlpha, F0, 1.0) * comp;
             ^
program_source:1565:20: warning: unused variable 'N' [-Wunused-variable]
            float3 N = normalize(vd.normalWorld);
                   ^

Shader source attached:
[simplehair.msl.txt](https://github.com/user-attachments/files/17319308/simplehair.msl.txt)

It appears to be related to two issues related to MSL and needs further investigation
1. Arrays need a size.
2. Arrays can't be passed by reference (?)

**Comment by ld-kerley** [comment]
Poking around a little, I think this boils down to two different situations where the GLSL syntax and MSL syntax are incompatible.

### Passing arrays to functions by reference
GLSL
```
void mx_hair_attenuation(float f, vec3 T, out vec3 Ap[4])
```
Currently the MSL shader generator mutates this to 
```
void mx_hair_attenuation(float f, vec3 T, thread vec3& Ap[4])
```
Which emits the following error from the Metal compiler.
```
program_source:1383:62: error: 'Ap' declared as array of references of type 'float3 &'
    void mx_hair_attenuation(float f, vec3 T, thread vec3& Ap[4])  // Ap
```

Instead the desired MSL code for this case is, actually just 
```
void mx_hair_attenuation(float f, vec3 T, thread vec3 Ap[4])
```
Which is a relatively simple addition to the `MslShaderGenerator::MetalizeGeneratedShader()` function - I can post the PR for this if we want to go this route.

### Initializing arrays
GLSL 
```
vec3 tint[4] = vec3[](tint_R, tint_TT, tint_TRT, tint_TRT);
```
Currently this code doesn't get mutated at all for the MSL case, and this raises the following error 
```
program_source:1448:29: error: expected expression
        vec3 tint[4] = vec3[](tint_R, tint_TT, tint_TRT, tint_TRT);
```
The MSL code to initialize an array is the same as C++
```
vec3 tint[4] = {tint_R, tint_TT, tint_TRT, tint_TRT};
```

This feels like a much harder case to robustly mutate the shader source code, especially when we consider potentially nested braces, etc. 

I'm wondering if this is another case where the more robust and pragmatic approach might be to create helper functions to perform this action in a language agnostic way, ie. (note uncompiled pseudo code)
```
vec3[] mx_createArray(vec3 a, vec3 b, vec3 c) 
{
 < language specific implementation >
}
```

Or I think we could just change the original GLSL code to initialize the array one element at a time, ie.
```
vec3 tint[4];
tint[0] = tint_R;
tint[1] = tint_TT;
tint[2] = tint_TRT;
tint[3] = tint_TRT;
```

**Comment by jstone-lucasfilm** [comment]
Thanks for this great analysis, @ld-kerley!

The idea of adding `mx_create_array` helper methods seems appealing, though we'd need to work through the implications for handling arbitrary data types and array lengths, without explicitly stating every permutation in our codebase.

If there's a direct approach along the lines of your last suggestion, where we initialize the array one element at a time, that seems like a good resolution as we're considering how best to generalize the handling of arrays.

**Comment by ld-kerley** [comment]
For the `mx_create_array()` case I was thinking of just taking the pragmatic approach of constructing just the function signatures that we need, currently two for Chiang hair - and just adding them as we need.

The only other pragmatic solution I can think of is to say we don't allow array initialization in GLSL/MSL for now - and just unfold it as per the last code snippet above. 

Honestly I think I'm leaning more towards the unfolding route for now - just to get us out of this bind.  The arrays in question are of very limited size, and one might hope that the GLSL/MSL compilers are smart enough to recognize any optimization lost by unfolding the initialization because of the locality of the code.

**Comment by jstone-lucasfilm** [comment]
Agred, @ld-kerley, and I don't even believe there's a performance decrease associated with unfolding the array code, given the way that a shader compiler would interpret this logic.

**Comment by jstone-lucasfilm** [comment]
Thanks to @ashwinbhat for this original report, and to @ld-kerley for the fix in #2057!
